File: packages/go/algorithms/device.go
package algorithms

import (
	"context"
	"crypto/md5"
	"fmt"
	"math"
	"sync"
	"time"

	fraudcatcher "github.com/fraud-catcher/go"
)

// DeviceConfig represents the configuration for the device algorithm
type DeviceConfig struct {
	EnableFingerprinting      bool    `json:"enable_fingerprinting"`
	SuspiciousDeviceThreshold float64 `json:"suspicious_device_threshold"`
	NewDeviceRiskMultiplier   float64 `json:"new_device_risk_multiplier"`
	DeviceVelocityWindow      int     `json:"device_velocity_window"`
	MaxDevicesPerUser         int     `json:"max_devices_per_user"`
	EnableDeviceReputation    bool    `json:"enable_device_reputation"`
	EnableDeviceClustering    bool    `json:"enable_device_clustering"`
	EnableCrossUserAnalysis   bool    `json:"enable_cross_user_analysis"`
	DeviceSharingThreshold    float64 `json:"device_sharing_threshold"`
}

// DeviceFingerprint represents a device fingerprint
type DeviceFingerprint struct {
	DeviceID         string                 `json:"device_id"`
	UserAgent        string                 `json:"user_agent"`
	IPAddress        string                 `json:"ip_address"`
	ScreenResolution string                 `json:"screen_resolution"`
	Timezone         string                 `json:"timezone"`
	Language         string                 `json:"language"`
	Platform         string                 `json:"platform"`
	FirstSeen        time.Time              `json:"first_seen"`
	LastSeen         time.Time              `json:"last_seen"`
	TransactionCount int                    `json:"transaction_count"`
	TotalAmount      float64                `json:"total_amount"`
	IsTrusted        bool                   `json:"is_trusted"`
	RiskScore        float64                `json:"risk_score"`
	Metadata         map[string]interface{} `json:"metadata"`
}

// DeviceProfile represents a user's device profile
type DeviceProfile struct {
	UserID            string                        `json:"user_id"`
	DeviceCount       int                           `json:"device_count"`
	PrimaryDevice     string                        `json:"primary_device"`
	DeviceHistory     map[string]*DeviceFingerprint `json:"device_history"`
	UserAgentPatterns map[string]int                `json:"user_agent_patterns"`
	IPAddressHistory  map[string]*IPData            `json:"ip_address_history"`
	LastDevice        string                        `json:"last_device"`
	LastTransaction   time.Time                     `json:"last_transaction"`
	LastUpdated       int64                         `json:"last_updated"`
}

// IPData represents data for an IP address
type IPData struct {
	IPAddress        string    `json:"ip_address"`
	FirstSeen        time.Time `json:"first_seen"`
	LastSeen         time.Time `json:"last_seen"`
	TransactionCount int       `json:"transaction_count"`
	TotalAmount      float64   `json:"total_amount"`
	IsTrusted        bool      `json:"is_trusted"`
}

// DeviceAlgorithm implements fraud detection based on device information
type DeviceAlgorithm struct {
	*fraudcatcher.BaseAlgorithm
	config             DeviceConfig
	deviceFingerprints map[string]*DeviceFingerprint
	userDevices        map[string]map[string]bool
	deviceUsers        map[string]map[string]bool
	userProfiles       map[string]*DeviceProfile
	mutex              sync.RWMutex
}

// NewDeviceAlgorithm creates a new device algorithm
func NewDeviceAlgorithm(config DeviceConfig) *DeviceAlgorithm {
	base := fraudcatcher.NewBaseAlgorithm("device", "Detects suspicious device patterns and device sharing", 0.5)

	algorithm := &DeviceAlgorithm{
		BaseAlgorithm:      base,
		config:             config,
		deviceFingerprints: make(map[string]*DeviceFingerprint),
		userDevices:        make(map[string]map[string]bool),
		deviceUsers:        make(map[string]map[string]bool),
		userProfiles:       make(map[string]*DeviceProfile),
	}

	return algorithm
}

// Analyze analyzes a transaction for device-based fraud
func (da *DeviceAlgorithm) Analyze(ctx context.Context, transaction fraudcatcher.Transaction, rule fraudcatcher.DetectionRule) (float64, error) {
	if !da.IsEnabled() {
		return 0.0, nil
	}

	if transaction.DeviceID == "" && transaction.UserAgent == "" && transaction.IPAddress == "" {
		return 0.0, nil // No device data available
	}

	deviceID := da.getOrCreateDeviceID(transaction)
	fingerprint := da.createFingerprint(transaction, deviceID)

	riskScore := 0.0

	// Check if device is new or suspicious
	existingFingerprint, exists := da.deviceFingerprints[deviceID]

	if !exists {
		// New device - check if user has too many devices
		userDeviceCount := da.getUserDeviceCount(transaction.UserID)
		if userDeviceCount >= da.config.MaxDevicesPerUser {
			riskScore += 0.6 // High risk for too many devices
		} else {
			riskScore += 0.3 * da.config.NewDeviceRiskMultiplier // Medium risk for new device
		}

		// Store new device
		da.deviceFingerprints[deviceID] = fingerprint
		da.addUserDevice(transaction.UserID, deviceID)
	} else {
		// Existing device - check for anomalies
		deviceRisk := da.calculateDeviceRisk(existingFingerprint, fingerprint)
		riskScore += deviceRisk

		// Update device fingerprint
		da.updateDeviceFingerprint(deviceID, transaction)
	}

	// Check device velocity (transactions per device)
	deviceVelocity := da.calculateDeviceVelocity(deviceID)
	if deviceVelocity > da.config.SuspiciousDeviceThreshold {
		riskScore += 0.4
	}

	// Check for device sharing patterns
	if da.config.EnableCrossUserAnalysis {
		sharingRisk := da.calculateDeviceSharingRisk(deviceID, transaction.UserID)
		riskScore += sharingRisk
	}

	// Update user profile
	da.updateUserProfile(transaction.UserID, deviceID, transaction)

	return math.Min(riskScore, 1.0), nil
}

// getOrCreateDeviceID gets or creates a device ID for the transaction
func (da *DeviceAlgorithm) getOrCreateDeviceID(transaction fraudcatcher.Transaction) string {
	if transaction.DeviceID != "" {
		return transaction.DeviceID
	}

	// Generate device ID from available data
	components := []string{
		transaction.UserAgent,
		transaction.IPAddress,
	}

	if transaction.Metadata != nil {
		if screenRes, exists := transaction.Metadata["screen_resolution"]; exists {
			components = append(components, fmt.Sprintf("%v", screenRes))
		}
	}

	hash := md5.Sum([]byte(fmt.Sprintf("%v", components)))
	return fmt.Sprintf("device_%x", hash)
}

// createFingerprint creates a device fingerprint from transaction data
func (da *DeviceAlgorithm) createFingerprint(transaction fraudcatcher.Transaction, deviceID string) *DeviceFingerprint {
	now := time.Now()

	fingerprint := &DeviceFingerprint{
		DeviceID:         deviceID,
		UserAgent:        transaction.UserAgent,
		IPAddress:        transaction.IPAddress,
		FirstSeen:        now,
		LastSeen:         now,
		TransactionCount: 1,
		TotalAmount:      transaction.Amount,
		IsTrusted:        false,
		RiskScore:        0.0,
		Metadata:         make(map[string]interface{}),
	}

	// Extract additional data from metadata
	if transaction.Metadata != nil {
		if screenRes, exists := transaction.Metadata["screen_resolution"]; exists {
			fingerprint.ScreenResolution = fmt.Sprintf("%v", screenRes)
		}
		if timezone, exists := transaction.Metadata["timezone"]; exists {
			fingerprint.Timezone = fmt.Sprintf("%v", timezone)
		}
		if language, exists := transaction.Metadata["language"]; exists {
			fingerprint.Language = fmt.Sprintf("%v", language)
		}
		if platform, exists := transaction.Metadata["platform"]; exists {
			fingerprint.Platform = fmt.Sprintf("%v", platform)
		}
	}

	return fingerprint
}

// calculateDeviceRisk calculates risk based on device fingerprint changes
func (da *DeviceAlgorithm) calculateDeviceRisk(existing, current *DeviceFingerprint) float64 {
	riskScore := 0.0

	// Check for device fingerprint changes
	if existing.UserAgent != current.UserAgent {
		riskScore += 0.3 // User agent changed
	}

	if existing.IPAddress != current.IPAddress {
		riskScore += 0.2 // IP address changed
	}

	if existing.ScreenResolution != current.ScreenResolution {
		riskScore += 0.1 // Screen resolution changed
	}

	if existing.Timezone != current.Timezone {
		riskScore += 0.1 // Timezone changed
	}

	// Check for rapid device changes
	timeDiff := current.LastSeen.Sub(existing.LastSeen)
	if timeDiff < time.Minute {
		riskScore += 0.2 // Rapid device switching
	}

	return math.Min(riskScore, 0.8)
}

// calculateDeviceVelocity calculates the velocity of transactions per device
func (da *DeviceAlgorithm) calculateDeviceVelocity(deviceID string) float64 {
	fingerprint, exists := da.deviceFingerprints[deviceID]
	if !exists {
		return 0
	}

	timeWindow := time.Duration(da.config.DeviceVelocityWindow) * time.Minute
	timeDiff := time.Since(fingerprint.FirstSeen)

	if timeDiff < timeWindow {
		return float64(fingerprint.TransactionCount) / timeDiff.Minutes() // Transactions per minute
	}

	return 0
}

// calculateDeviceSharingRisk calculates risk based on device sharing patterns
func (da *DeviceAlgorithm) calculateDeviceSharingRisk(deviceID, userID string) float64 {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	deviceUsers, exists := da.deviceUsers[deviceID]
	if !exists {
		return 0.0
	}

	userCount := len(deviceUsers)
	if userCount > 1 {
		// Device is used by multiple users
		sharingRatio := float64(userCount-1) / float64(userCount)
		if sharingRatio > da.config.DeviceSharingThreshold {
			return 0.5 // High risk for device sharing
		}
		return 0.2 // Medium risk for device sharing
	}

	return 0.0
}

// getUserDeviceCount returns the number of devices for a user
func (da *DeviceAlgorithm) getUserDeviceCount(userID string) int {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	devices, exists := da.userDevices[userID]
	if !exists {
		return 0
	}

	return len(devices)
}

// addUserDevice adds a device to a user's device list
func (da *DeviceAlgorithm) addUserDevice(userID, deviceID string) {
	da.mutex.Lock()
	defer da.mutex.Unlock()

	if da.userDevices[userID] == nil {
		da.userDevices[userID] = make(map[string]bool)
	}
	da.userDevices[userID][deviceID] = true

	if da.deviceUsers[deviceID] == nil {
		da.deviceUsers[deviceID] = make(map[string]bool)
	}
	da.deviceUsers[deviceID][userID] = true
}

// updateDeviceFingerprint updates a device fingerprint with new transaction data
func (da *DeviceAlgorithm) updateDeviceFingerprint(deviceID string, transaction fraudcatcher.Transaction) {
	da.mutex.Lock()
	defer da.mutex.Unlock()

	fingerprint, exists := da.deviceFingerprints[deviceID]
	if exists {
		fingerprint.LastSeen = transaction.Timestamp
		fingerprint.TransactionCount++
		fingerprint.TotalAmount += transaction.Amount
	}
}

// updateUserProfile updates the user's device profile
func (da *DeviceAlgorithm) updateUserProfile(userID, deviceID string, transaction fraudcatcher.Transaction) {
	da.mutex.Lock()
	defer da.mutex.Unlock()

	profile, exists := da.userProfiles[userID]
	if !exists {
		profile = &DeviceProfile{
			UserID:            userID,
			DeviceHistory:     make(map[string]*DeviceFingerprint),
			UserAgentPatterns: make(map[string]int),
			IPAddressHistory:  make(map[string]*IPData),
			LastUpdated:       0,
		}
		da.userProfiles[userID] = profile
	}

	profile.DeviceCount = da.getUserDeviceCount(userID)
	profile.LastDevice = deviceID
	profile.LastTransaction = transaction.Timestamp
	profile.LastUpdated = transaction.Timestamp.Unix()

	// Update user agent patterns
	if transaction.UserAgent != "" {
		profile.UserAgentPatterns[transaction.UserAgent]++
	}

	// Update IP address history
	if transaction.IPAddress != "" {
		if ipData, exists := profile.IPAddressHistory[transaction.IPAddress]; exists {
			ipData.LastSeen = transaction.Timestamp
			ipData.TransactionCount++
			ipData.TotalAmount += transaction.Amount
		} else {
			profile.IPAddressHistory[transaction.IPAddress] = &IPData{
				IPAddress:        transaction.IPAddress,
				FirstSeen:        transaction.Timestamp,
				LastSeen:         transaction.Timestamp,
				TransactionCount: 1,
				TotalAmount:      transaction.Amount,
				IsTrusted:        false,
			}
		}
	}

	// Set primary device if not set
	if profile.PrimaryDevice == "" {
		profile.PrimaryDevice = deviceID
	}
}

// GetDeviceFingerprint returns a device fingerprint by ID
func (da *DeviceAlgorithm) GetDeviceFingerprint(deviceID string) *DeviceFingerprint {
	da.mutex.RLock()
	defer da.mutex.RUnlock()
	return da.deviceFingerprints[deviceID]
}

// GetUserDevices returns the devices for a user
func (da *DeviceAlgorithm) GetUserDevices(userID string) []string {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	devices, exists := da.userDevices[userID]
	if !exists {
		return nil
	}

	var deviceList []string
	for deviceID := range devices {
		deviceList = append(deviceList, deviceID)
	}

	return deviceList
}

// MarkDeviceAsTrusted marks a device as trusted
func (da *DeviceAlgorithm) MarkDeviceAsTrusted(deviceID string) {
	da.mutex.Lock()
	defer da.mutex.Unlock()

	fingerprint, exists := da.deviceFingerprints[deviceID]
	if exists {
		fingerprint.IsTrusted = true
	}
}

// GetDeviceStats returns statistics for a device
func (da *DeviceAlgorithm) GetDeviceStats(deviceID string) map[string]interface{} {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	fingerprint, exists := da.deviceFingerprints[deviceID]
	if !exists {
		return map[string]interface{}{
			"transaction_count": 0,
			"total_amount":      0.0,
			"is_trusted":        false,
		}
	}

	return map[string]interface{}{
		"transaction_count": fingerprint.TransactionCount,
		"total_amount":      fingerprint.TotalAmount,
		"is_trusted":        fingerprint.IsTrusted,
		"first_seen":        fingerprint.FirstSeen,
		"last_seen":         fingerprint.LastSeen,
		"risk_score":        fingerprint.RiskScore,
	}
}

// GetUserProfile returns the user profile for a given user ID
func (da *DeviceAlgorithm) GetUserProfile(userID string) *DeviceProfile {
	da.mutex.RLock()
	defer da.mutex.RUnlock()
	return da.userProfiles[userID]
}

// GetDeviceUsers returns the users associated with a device
func (da *DeviceAlgorithm) GetDeviceUsers(deviceID string) []string {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	users, exists := da.deviceUsers[deviceID]
	if !exists {
		return nil
	}

	var userList []string
	for userID := range users {
		userList = append(userList, userID)
	}

	return userList
}

// GetStats returns statistics for the device algorithm
func (da *DeviceAlgorithm) GetStats() map[string]interface{} {
	da.mutex.RLock()
	defer da.mutex.RUnlock()

	stats := da.BaseAlgorithm.GetStats()
	stats["device_fingerprints_count"] = len(da.deviceFingerprints)
	stats["user_profiles_count"] = len(da.userProfiles)
	stats["total_devices"] = len(da.deviceFingerprints)
	stats["config"] = da.config

	return stats
}

// Reset resets the algorithm state
func (da *DeviceAlgorithm) Reset() error {
	da.mutex.Lock()
	defer da.mutex.Unlock()

	da.deviceFingerprints = make(map[string]*DeviceFingerprint)
	da.userDevices = make(map[string]map[string]bool)
	da.deviceUsers = make(map[string]map[string]bool)
	da.userProfiles = make(map[string]*DeviceProfile)

	return nil
}

File: packages/python/src/fraud_catcher/algorithms/device.py
"""
Device-based fraud detection algorithm.
"""

import time
from datetime import datetime
from typing import Dict, List, Optional, Set, Any
from dataclasses import dataclass
from ..models import Transaction, DetectionRule


@dataclass
class DeviceConfig:
    """Configuration for device algorithm."""
    enable_fingerprinting: bool
    suspicious_device_threshold: int
    new_device_risk_multiplier: float
    device_velocity_window: int  # in minutes
    max_devices_per_user: int


@dataclass
class DeviceFingerprint:
    """Device fingerprint information."""
    device_id: str
    user_agent: str
    ip_address: str
    screen_resolution: Optional[str] = None
    timezone: Optional[str] = None
    language: Optional[str] = None
    platform: Optional[str] = None
    first_seen: Optional[datetime] = None
    last_seen: Optional[datetime] = None
    transaction_count: int = 0
    total_amount: float = 0.0
    is_trusted: bool = False


class DeviceAlgorithm:
    """Detects fraud based on device patterns and sharing."""
    
    def __init__(self, config: DeviceConfig):
        self.config = config
        self.device_fingerprints: Dict[str, DeviceFingerprint] = {}
        self.user_devices: Dict[str, Set[str]] = {}
    
    async def analyze(self, transaction: Transaction, rule: DetectionRule) -> float:
        """Analyze transaction for device-based fraud patterns."""
        if not transaction.device_id and not transaction.user_agent and not transaction.ip_address:
            return 0.0  # No device data available
        
        device_id = transaction.device_id or self._generate_device_id(transaction)
        fingerprint = self._create_fingerprint(transaction, device_id)
        
        risk_score = 0.0
        
        # Check if device is new or suspicious
        existing_fingerprint = self.device_fingerprints.get(device_id)
        
        if not existing_fingerprint:
            # New device - check if user has too many devices
            user_device_count = self._get_user_device_count(transaction.user_id)
            if user_device_count >= self.config.max_devices_per_user:
                risk_score += 0.6  # High risk for too many devices
            else:
                risk_score += 0.3  # Medium risk for new device
            
            # Store new device
            self.device_fingerprints[device_id] = fingerprint
            self._add_user_device(transaction.user_id, device_id)
        else:
            # Existing device - check for anomalies
            device_risk = self._calculate_device_risk(existing_fingerprint, fingerprint)
            risk_score += device_risk
            
            # Update device fingerprint
            self._update_device_fingerprint(device_id, transaction)
        
        # Check device velocity (transactions per device)
        device_velocity = self._calculate_device_velocity(device_id)
        if device_velocity > self.config.suspicious_device_threshold:
            risk_score += 0.4
        
        # Check for device sharing patterns
        sharing_risk = self._calculate_device_sharing_risk(device_id, transaction.user_id)
        risk_score += sharing_risk
        
        return min(risk_score, 1.0)
    
    def _generate_device_id(self, transaction: Transaction) -> str:
        """Generate device ID from available data."""
        components = [
            transaction.user_agent or 'unknown',
            transaction.ip_address or 'unknown',
            transaction.metadata.get('screen_resolution') if transaction.metadata else 'unknown'
        ]
        
        return f"device_{hash('|'.join(components)) % 1000000}"
    
    def _create_fingerprint(self, transaction: Transaction, device_id: str) -> DeviceFingerprint:
        """Create device fingerprint from transaction."""
        now = datetime.now()
        
        return DeviceFingerprint(
            device_id=device_id,
            user_agent=transaction.user_agent or '',
            ip_address=transaction.ip_address or '',
            screen_resolution=transaction.metadata.get('screen_resolution') if transaction.metadata else None,
            timezone=transaction.metadata.get('timezone') if transaction.metadata else None,
            language=transaction.metadata.get('language') if transaction.metadata else None,
            platform=transaction.metadata.get('platform') if transaction.metadata else None,
            first_seen=now,
            last_seen=now,
            transaction_count=1,
            total_amount=transaction.amount,
            is_trusted=False
        )
    
    def _calculate_device_risk(self, existing: DeviceFingerprint, current: DeviceFingerprint) -> float:
        """Calculate risk based on device fingerprint changes."""
        risk_score = 0.0
        
        # Check for device fingerprint changes
        if existing.user_agent != current.user_agent:
            risk_score += 0.3  # User agent changed
        
        if existing.ip_address != current.ip_address:
            risk_score += 0.2  # IP address changed
        
        if existing.screen_resolution != current.screen_resolution:
            risk_score += 0.1  # Screen resolution changed
        
        if existing.timezone != current.timezone:
            risk_score += 0.1  # Timezone changed
        
        # Check for rapid device changes
        if existing.last_seen and current.last_seen:
            time_diff = (current.last_seen - existing.last_seen).total_seconds()
            if time_diff < 60:  # Less than 1 minute
                risk_score += 0.2  # Rapid device switching
        
        return min(risk_score, 0.8)
    
    def _calculate_device_velocity(self, device_id: str) -> float:
        """Calculate device velocity (transactions per minute)."""
        fingerprint = self.device_fingerprints.get(device_id)
        if not fingerprint:
            return 0.0
        
        time_window_seconds = self.config.device_velocity_window * 60
        now = datetime.now()
        
        if fingerprint.first_seen:
            time_diff = (now - fingerprint.first_seen).total_seconds()
            if time_diff < time_window_seconds:
                return fingerprint.transaction_count / (time_diff / 60)  # Transactions per minute
        
        return 0.0
    
    def _calculate_device_sharing_risk(self, device_id: str, user_id: str) -> float:
        """Calculate risk of device sharing."""
        # Check if device is used by multiple users
        device_users = self._get_device_users(device_id)
        if len(device_users) > 1:
            return 0.5  # High risk for device sharing
        
        return 0.0
    
    def _get_user_device_count(self, user_id: str) -> int:
        """Get number of devices for user."""
        return len(self.user_devices.get(user_id, set()))
    
    def _add_user_device(self, user_id: str, device_id: str) -> None:
        """Add device to user's device list."""
        if user_id not in self.user_devices:
            self.user_devices[user_id] = set()
        self.user_devices[user_id].add(device_id)
    
    def _get_device_users(self, device_id: str) -> Set[str]:
        """Get users associated with device."""
        users = set()
        for user_id, devices in self.user_devices.items():
            if device_id in devices:
                users.add(user_id)
        return users
    
    def _update_device_fingerprint(self, device_id: str, transaction: Transaction) -> None:
        """Update device fingerprint with new transaction."""
        fingerprint = self.device_fingerprints.get(device_id)
        if fingerprint:
            fingerprint.last_seen = datetime.now()
            fingerprint.transaction_count += 1
            fingerprint.total_amount += transaction.amount
    
    # Utility methods
    def get_device_fingerprint(self, device_id: str) -> Optional[DeviceFingerprint]:
        """Get device fingerprint by ID."""
        return self.device_fingerprints.get(device_id)
    
    def get_user_devices(self, user_id: str) -> List[str]:
        """Get list of device IDs for user."""
        return list(self.user_devices.get(user_id, set()))
    
    def mark_device_as_trusted(self, device_id: str) -> None:
        """Mark device as trusted."""
        fingerprint = self.device_fingerprints.get(device_id)
        if fingerprint:
            fingerprint.is_trusted = True
    
    def get_device_stats(self, device_id: str) -> Dict[str, Any]:
        """Get device statistics."""
        fingerprint = self.device_fingerprints.get(device_id)
        if not fingerprint:
            return {'transaction_count': 0, 'total_amount': 0.0, 'is_trusted': False}
        
        return {
            'transaction_count': fingerprint.transaction_count,
            'total_amount': fingerprint.total_amount,
            'is_trusted': fingerprint.is_trusted
        }
